File: libgomp/testsuite/libgomp.c-c++-common/get-mapped-ptr-3.c
#include <omp.h>
#include <stdlib.h>

int
main ()
{
  int d = omp_get_default_device ();
  int id = omp_get_initial_device ();
  int a[0];
  int b[] = { 24, 42 };
  void *p1 = NULL, *p2 = NULL;

  if (d < 0 || d >= omp_get_num_devices ())
    d = id;

  void *p = omp_target_alloc (sizeof (int), d);
  if (p == NULL)
    return 0;

  if (omp_target_associate_ptr (a, p, sizeof (int), 0, d) != 0)
    return 0;

  if (omp_get_mapped_ptr (a, d) != (d == id ? a : p))
    abort ();

  if (omp_target_disassociate_ptr (a, d) != 0)
    abort ();

  if (omp_get_mapped_ptr (a, d) != (d == id ? a : NULL))
    abort ();

  #pragma omp target data map(alloc: a, b[1:0]) device(d)
  {
    #pragma omp target map(from: p1, p2) map(alloc: a, b[1:0]) device(d)
    {
      p1 = &a;
      p2 = &b[1];
    }

    /* This is probably expected to be p1/p2 instead of NULL. Zero-length arrays
       as list items of the map clause are currently not inserted into the mem
       map ?! However by returning NULL, omp_get_mapped_ptr is consistent with
       omp_target_is_present.  */
    if (omp_get_mapped_ptr (a, d) != NULL
	|| omp_get_mapped_ptr (&b[1], d) != NULL)
      abort ();
  }

  omp_target_free (p, d);
  return 0;
}

File: libgomp/testsuite/libgomp.fortran/get-mapped-ptr-3.f90
program main
  use omp_lib
  use iso_c_binding
  implicit none (external, type)
  integer :: d, id
  type(c_ptr) :: p, p1, p2
  integer, target :: a(1:0), b(1:2)

  d = omp_get_default_device ()
  id = omp_get_initial_device ()

  if (d < 0 .or. d >= omp_get_num_devices ()) &
    d = id

  p = omp_target_alloc (c_sizeof (c_int), d)
  if (.not. c_associated (p)) &
    stop 0  ! okay

  if (omp_target_associate_ptr (c_loc (a), p, c_sizeof (c_int), &
                                0_c_size_t, d) == 0) then

  if(.not. c_associated (omp_get_mapped_ptr (c_loc (a), d), p)) &
    stop 1

  if (omp_target_disassociate_ptr (c_loc (a), d) /= 0) &
    stop 2

  if(c_associated (omp_get_mapped_ptr (c_loc (a), d))) &
    stop 3

  !$omp target data map(alloc: a) device(d)
    !$omp target map(from: p1) map(alloc: a) device(d)
    p1 = c_loc (a);
    !$omp end target
    if (c_associated (omp_get_mapped_ptr (c_loc (a), d))) &
      stop 4
  !$omp end target data

  !$omp target data map(alloc: b(1:0)) device(d)
    !$omp target map(from: p2) map(alloc: b(1:0)) device(d)
    p2 = c_loc (b(1));
    !$omp end target
    if (c_associated (omp_get_mapped_ptr (c_loc (b(1)), d))) &
      stop 5
  !$omp end target data
  end if
  call omp_target_free (p, d)
end program main
